engine/src/main/java/org/terasology/engine/world/chunks/pipeline/ChunkProcessingPipeline.java

```diff
@@ -45,119 +31,149 @@
-        return (o1, o2) -> {
-                Object unwrapped1 = ReflectionUtil.readField(o1, "task");
-                Object unwrapped2 = ReflectionUtil.readField(o2, "task");
-                return comparable.compare((Future<Chunk>) unwrapped1, (Future<Chunk>) unwrapped2);
-        };
+    public void notifyUpdate() {
+        synchronized (waitForNewChunks) {
+            waitForNewChunks.notifyAll();
+        }
     }
 
-    /**
-     * Reactor thread. Handles all ChunkTask dependency logic and running.
-     */
-    private void chunkTaskHandler() {
+    private void stopPoolThreads() {
+        shouldStop.set(true);
+        notifyUpdate();
+    }
+
+    private void runPoolThread() {
+        Preconditions.checkState(!stages.isEmpty(), "ChunkProcessingPipeline must to have at least one stage");
+
         try {
-            while (!executor.isTerminated()) {
-                PositionFuture<Chunk> future = (PositionFuture<Chunk>) chunkProcessor.take();
-                ChunkProcessingInfo chunkProcessingInfo = chunkProcessingInfoMap.get(future.getPosition());
-                if (chunkProcessingInfo == null) {
-                    continue; // chunk processing was cancelled.
+            while (!shouldStop.get()) {
+                // Prioritize advancing generation of chunks that were already started
+                if (processChunkTasks()) {
+                    continue;
+                }
+
+                // But if there aren't any chunks that can advance to the next stage right now, start some new chunks
+                for (int i = 0; i < NUM_CHUNKS_AT_ONCE; i++) {
+                    Chunk chunk = null;
+                    if (initialChunkProvider.hasNext()) {
+                        chunk = initialChunkProvider.next(chunkProcessingInfoMap.keySet());
+                    }
+                    if (chunk == null) {
+                        if (shouldStop.get()) {
+                            break;
+                        }
+                        synchronized (waitForNewChunks) {
+                            waitForNewChunks.wait();
+                        }
+                        continue;
+                    }
+                    Vector3ic position = chunk.getPosition();
+
+                    ChunkProcessingInfo chunkProcessingInfo = new ChunkProcessingInfo(position);
+                    chunkProcessingInfo.setChunk(chunk);
+                    chunkProcessingInfo.nextStage(stages);
+                    chunkProcessingInfo.makeChunkTask();
+                    chunkProcessingInfoMap.put(position, chunkProcessingInfo);
                 }
-                onStageDone(future, chunkProcessingInfo);
             }
         } catch (InterruptedException e) {
-            if (!executor.isTerminated()) {
-                logger.error("Reactor thread was interrupted", e);
-            }
-            reactor.interrupt();
+            Thread.currentThread().interrupt();
```

Comment:
This should actually never happen now that it doesn't interrupt the worker threads to stop them anymore, but it used to be there for them to shut down gracefully.

Reply:
What was the problem with using interrupt on the worker threads? That sounds like it was working as intended?

Reply:
It would sometimes interrupt the threads while they were doing IO, which would at least cause some scary error messages. It wasn't actually simpler, and now that the waiting logic is correct, it's not needed anyway.